app/main/routes.py

```diff
@@ -49,55 +49,63 @@ def upload():
 
     if request.method == "POST":
         excel_file = request.files.get("ingest_spreadsheet")
-        error = check_file(excel_file)
-        if error:
-            current_app.logger.info(
-                PRE_VALIDATION_ERROR_LOG.format(error=error) if Config.ENABLE_VALIDATION_LOGGING else PRE_VALIDATION_LOG
-            )
-            return render_template(
-                "upload.html",
-                pre_error=[error],
-                days_to_deadline=calculate_days_to_deadline(),
-                reporting_period=Config.REPORTING_PERIOD,
-                fund=Config.FUND_NAME,
-            )
 
-        success, pre_errors, validation_errors, metadata = post_ingest(
-            excel_file,
-            {"reporting_round": 4, "auth": json.dumps(g.auth.get_auth_dict()), "do_load": is_load_enabled()},
-        )
+        if file_errors := check_file(excel_file):
+            pre_errors, validation_errors, metadata = file_errors, None, None
+        else:
+            pre_errors, validation_errors, metadata = post_ingest(
+                excel_file,
+                {"reporting_round": 4, "auth": json.dumps(g.auth.get_auth_dict()), "do_load": is_load_enabled()},
+            )
 
-        if success:
-            if Config.SEND_CONFIRMATION_EMAILS:
-                send_confirmation_emails(excel_file, metadata, user_email=g.user.email)
-            metadata["User"] = g.user.email
-            current_app.logger.info(f"Upload successful: {metadata}")
-            return render_template("success.html", file_name=excel_file.filename)
-        elif pre_errors:
+        if pre_errors:
+            # Pre-validation failure
             if Config.ENABLE_VALIDATION_LOGGING:
                 for pre_err in pre_errors:
                     current_app.logger.info(PRE_VALIDATION_ERROR_LOG.format(error=pre_err))
             else:
                 current_app.logger.info(PRE_VALIDATION_LOG)
 
+            if file_errors and "validation_errors" in session:
+                return render_template(
+                    "validation-errors.html",
+                    pre_error=pre_errors,
+                    validation_errors=session.get("validation_errors"),
+                )
+
+            if "validation_errors" in session:
+                session.pop("validation_errors")
             return render_template(
                 "upload.html",
                 pre_error=pre_errors,
                 days_to_deadline=calculate_days_to_deadline(),
                 reporting_period=Config.REPORTING_PERIOD,
                 fund=Config.FUND_NAME,
             )
```

Comment:
I think it's clearer if we just use pop with a default value and then use the result through the rest like:

```suggestion
            validation_errors = session.pop("validation_errors", None)

            if file_errors and validation_errors:
                return render_template(
                    "validation-errors.html",
                    pre_error=pre_errors,
                    validation_errors=validation_errors,
                )

            return render_template(
                "upload.html",
                pre_error=pre_errors,
                days_to_deadline=calculate_days_to_deadline(),
                reporting_period=Config.REPORTING_PERIOD,
                fund=Config.FUND_NAME,
            )
        
```